File: app/src/main/java/com/ne1c/developerstalk/api/GitterApi.java
package com.ne1c.developerstalk.api;

import com.ne1c.developerstalk.models.AuthResponseModel;
import com.ne1c.developerstalk.models.MessageModel;
import com.ne1c.developerstalk.models.RoomModel;
import com.ne1c.developerstalk.models.UserModel;

import java.util.ArrayList;

import okhttp3.ResponseBody;
import retrofit2.http.DELETE;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;
import retrofit2.http.Streaming;
import rx.Observable;

public interface GitterApi {

    @GET("/v1/rooms")
    Observable<ArrayList<RoomModel>> getCurrentUserRooms(@Header("Authorization") String access_token);

    @GET("/v1/rooms/{roomId}/chatMessages")
    Observable<ArrayList<MessageModel>> getMessagesRoom(@Header("Authorization") String access_token,
                                                        @Path("roomId") String roomId,
                                                        @Query("limit") int limit);

    @FormUrlEncoded
    @POST("/v1/rooms/{roomId}/chatMessages")
    Observable<MessageModel> sendMessage(@Header("Authorization") String access_token,
                                         @Path("roomId") String roomId,
                                         @Field("text") String text);

    @FormUrlEncoded
    @PUT("/v1/rooms/{roomId}/chatMessages/{chatMessageId}")
    Observable<MessageModel> updateMessage(@Header("Authorization") String access_token,
                                           @Path("roomId") String roomId,
                                           @Path("chatMessageId") String chatMessageId,
                                           @Field("text") String messageText);

    @GET("/v1/user")
    Observable<ArrayList<UserModel>> getCurrentUser(@Header("Authorization") String access_token);

    @GET("/v1/rooms/{roomId}/chatMessages")
    Observable<ArrayList<MessageModel>> getMessagesBeforeId(@Header("Authorization") String access_token,
                                                            @Path("roomId") String roomId,
                                                            @Query("limit") int limit,
                                                            @Query("beforeId") String beforeId);

    @FormUrlEncoded
    @POST("/login/oauth/token")
    Observable<AuthResponseModel> authorization(@Field("client_id") String client_id,
                                                @Field("client_secret") String client_secret,
                                                @Field("code") String code,
                                                @Field("grant_type") String grant_type,
                                                @Field("redirect_uri") String redirect_uri);

    @FormUrlEncoded
    @POST("/v1/user/{userId}/rooms/{roomId}/unreadItems")
    Observable<ResponseBody> readMessages(@Header("Authorization") String access_token, @Path("userId") String userId,
                                          @Path("roomId") String roomId, @Field("chat") String[] chat);

    @DELETE("/v1/rooms/{roomId}/users/{userId}")
    Observable<ResponseBody> leaveRoom(@Header("Authorization") String access_token,
                                   @Path("roomId") String roomId,
                                   @Path("userId") String userId);

    @POST("/v1/rooms")
    @FormUrlEncoded
    Observable<RoomModel> joinRoom(@Header("Authorization") String access_token,
                                   @Field("uri") String roomUri);

    @GET("/v1/rooms")
    Observable<ResponseBody> searchRooms(@Header("Authorization") String access_token,
                                     @Query("q") String searchTerm);

    @GET("/v1/rooms")
    Observable<ResponseBody> searchRooms(@Header("Authorization") String access_token,
                                     @Query("q") String searchTerm,
                                     @Query("limit") int limit);

    @GET("/v1/user")
    Observable<ResponseBody> searchUsers(@Header("Authorization") String access_token,
                                     @Query("q") String searchTerm);

    @Streaming
    @GET("/v1/rooms/{roomId}/chatMessages")
    Observable<ResponseBody> getRoomStream(@Header("Authorization") String access_token,
                                       @Path("roomId") String roomId);

    @POST("/v1/rooms")
    Observable<ArrayList<RoomModel>> getSearchableRooms(@Header("Authorization") String access_token,
                                                        @Field("q") String query);
}

File: app/src/main/kotlin/com/ne1c/gitteroid/api/GitterApi.kt
package com.ne1c.gitteroid.api

import com.ne1c.gitteroid.api.responses.JoinRoomResponse
import com.ne1c.gitteroid.api.responses.StatusResponse
import com.ne1c.gitteroid.models.data.*
import okhttp3.ResponseBody
import retrofit2.http.*
import rx.Observable
import java.util.*

interface GitterApi {
    @GET("/v1/rooms")
    fun getCurrentUserRooms(@Header("Authorization") access_token: String): Observable<ArrayList<RoomModel>>

    @GET("/v1/rooms/{roomId}/chatMessages")
    fun getMessagesRoom(@Header("Authorization") access_token: String,
                        @Path("roomId") roomId: String,
                        @Query("limit") limit: Int): Observable<ArrayList<MessageModel>>

    @FormUrlEncoded
    @POST("/v1/rooms/{roomId}/chatMessages")
    fun sendMessage(@Header("Authorization") access_token: String,
                    @Path("roomId") roomId: String,
                    @Field("text") text: String): Observable<MessageModel>

    @FormUrlEncoded
    @PUT("/v1/rooms/{roomId}/chatMessages/{chatMessageId}")
    fun updateMessage(@Header("Authorization") access_token: String,
                      @Path("roomId") roomId: String,
                      @Path("chatMessageId") chatMessageId: String,
                      @Field("text") messageText: String): Observable<MessageModel>

    @GET("/v1/user")
    fun getCurrentUser(@Header("Authorization") access_token: String): Observable<ArrayList<UserModel>>

    @GET("/v1/rooms/{roomId}/chatMessages")
    fun getMessagesBeforeId(@Header("Authorization") access_token: String,
                            @Path("roomId") roomId: String,
                            @Query("limit") limit: Int,
                            @Query("beforeId") beforeId: String): Observable<ArrayList<MessageModel>>

    @FormUrlEncoded
    @POST
    fun     authorization(@Url authUrl: String,
                      @Field("client_id") client_id: String,
                      @Field("client_secret") client_secret: String,
                      @Field("code") code: String,
                      @Field("grant_type") grant_type: String,
                      @Field("redirect_uri") redirect_uri: String): Observable<AuthResponseModel>

    @FormUrlEncoded
    @POST("/v1/user/{userId}/rooms/{roomId}/unreadItems")
    fun readMessages(@Header("Authorization") access_token: String,
                     @Path("userId") userId: String,
                     @Path("roomId") roomId: String,
                     @Field("chat") chat: Array<String>): Observable<StatusResponse>

    @DELETE("/v1/rooms/{roomId}/users/{userId}")
    fun leaveRoom(@Header("Authorization") access_token: String,
                  @Path("roomId") roomId: String,
                  @Path("userId") userId: String): Observable<StatusResponse>

    @POST("/v1/rooms")
    @FormUrlEncoded
    fun joinRoom(@Header("Authorization") access_token: String,
                 @Field("uri") roomUri: String): Observable<JoinRoomResponse>

    @GET("/v1/user")
    fun searchUsers(@Header("Authorization") access_token: String,
                    @Query("q") searchTerm: String): Observable<ResponseBody>

    @GET("/v1/rooms")
    fun searchRooms(@Header("Authorization") access_token: String,
                    @Query("q") query: String,
                    @Query("limit") limit: Int,
                    @Query("offset") offset: Int): Observable<SearchRoomsResponse>  // Offset not working
}